Porting of the X-Ray Latent Diffusion Models model from MONAI Generative Models into model-zoo.

virginiafdez opened the pull request:
### Description 

As part of the porting of MONAI Generative Models, the models present in the model zoo from that repo (https://github.com/Project-MONAI/GenerativeModels/tree/main/model-zoo) need to be ported into the main model zoo from MONAI.
This is one of the models, a latent diffusion model that generates X-Rays.

### Status
**Ready

### Please ensure all the checkboxes:
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Codeformat tests passed locally by running `./runtests.sh --codeformat`.
- [x] In-line docstrings updated.
- [x ] Please ensure the naming rules in config files meet our requirements (please refer to: `CONTRIBUTING.md`).
- [x] Ensure versions of packages such as `monai`, `pytorch` and `numpy` are correct in `metadata.json`.
- [x] Descriptions should be consistent with the content, such as `eval_metrics` of the provided weights and TorchScript modules.
- [x] Files larger than 25MB are excluded and replaced by providing download links in `large_file.yml`.
- [x] Avoid using path that contains personal information within config files (such as use `/home/your_name/` for `"bundle_root"`).

Commit:
Initial commit of CXR Latent Diffusion model from the GenerativeModels repo.

Signed-off-by: Virginia <virginia.fernandez@kcl.ac.uk>

Commit:
Modify the README.md file as it contained something from the old repo that no longer applies.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ericspod commented on models/cxr_image_synthesis_latent_diffusion_model/configs/inference.json:
```diff
@@ -0,0 +1,108 @@
+        "cross_attention_dim": 1024
+    },
+    "load_diffusion_path": "$@model_dir + '/diffusion_model.pth'",
+    "load_diffusion": "$@diffusion_def.load_old_state_dict(torch.load(@load_diffusion_path)) if bool(@load_old) else @diffusion_def.load_state_dict(torch.load(@load_diffusion_path))",
```
```suggestion
    "load_diffusion_func": "$@diffusion_def.load_old_state_dict if bool(@load_old) else @diffusion_def.load_state_dict",
    "load_diffusion": "$@load_diffusion_func(torch.load(@load_diffusion_path))",
```
If you wanted to reduce the length of this line you can break things up into two values like this.

ericspod commented on models/cxr_image_synthesis_latent_diffusion_model/configs/metadata.json:
```diff
@@ -0,0 +1,73 @@
+    "numpy_version": "1.22.4",
+    "optional_packages_version": {
+        "nibabel": "4.0.1",
+        "generative": "0.1.0",
```
```suggestion
```

ericspod commented on models/cxr_image_synthesis_latent_diffusion_model/configs/inference.json:
```diff
@@ -0,0 +1,108 @@
+        "_requires_": "@create_output_dir",
+        "output_dir": "@output_dir"
+    },
+    "save_jpg": "$@saver.save(@sample, @out_file)",
```
This and `save` are the entry points essentially for this "script". Although the workflow engine classes aren't used here, we should still stick the pattern in other bundles where the entry point is called `run`, such as [here](https://github.com/Project-MONAI/model-zoo/blob/d48630661c001a642a4704abdd7db4e3a722b397/models/spleen_ct_segmentation/configs/evaluate.json#L73).

ericspod commented on models/cxr_image_synthesis_latent_diffusion_model/docs/README.md:
````diff
@@ -0,0 +1,71 @@
+Examples:
+
+```shell
+$ python -m monai.bundle run save_jpg --config_file configs/inference.json --prompt "Big right-sided pleural effusion" --guidance_scale 7.0
````
If the entry point is named `run` instead of `save_jpg`, this name can be omitted from the command lines.

Commit:
Modification of metadata to remove generative package requirement, no longer necessary. On inference, modification of the entry point from save_jpg to run to allow for run to be called alone, and truncation of loading statements which are too long. Modification of the README to include command example changes.

virginiafdez committed:
Merge branch 'Project-MONAI:dev' into xray_classifier_port

Commit:
Merge branch 'xray_classifier_port' of https://github.com/virginiafdez/model-zoo into xray_classifier_port

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

KumoLiu reviewed (commented):
Thanks for the contribution. Would be great to add corresponding test here:
https://github.com/Project-MONAI/model-zoo/tree/dev/ci/unit_tests

KumoLiu commented on models/cxr_image_synthesis_latent_diffusion_model/large_files.yml:
```diff
@@ -0,0 +1,9 @@
+large_files:
+  - path: "models/autoencoder.pth"
+    url: "https://drive.google.com/uc?export=download&id=1paDN1m-Q_Oy8d_BanPkRTi3RlNB_Sv_h"
+    hash_val: ""
```
Please also include the hash val here.

KumoLiu commented on models/cxr_image_synthesis_latent_diffusion_model/configs/metadata.json:
```diff
@@ -0,0 +1,73 @@
+    "schema": "https://github.com/Project-MONAI/MONAI-extra-test-data/releases/download/0.8.1/meta_schema_20220324.json",
+    "version": "1.0.0",
+    "changelog": {
+        "0.2": "Flipped images fixed"
```
Should also include "1.0.0" here?

Commit:
Added hash values and types to large_files.yml.
For metadata, adjusted the versions, setting current to Initial Release. I also removed nibabel from the optional package requirements as this CXR produces JPG and does not deal with nifti files.

Commit:
Merge branch 'xray_classifier_port' of https://github.com/virginiafdez/model-zoo into xray_classifier_port

yiheng-wang-nv commented on models/cxr_image_synthesis_latent_diffusion_model/configs/inference.json:
```diff
@@ -0,0 +1,110 @@
+{
+    "imports": [
```
Hi @virginiafdez , thanks for the PR. Could you also help to add a test file? The purpose is to ensure the bundle is runnable, and here is an example:
https://github.com/Project-MONAI/model-zoo/blob/dev/ci/unit_tests/test_spleen_ct_segmentation.py#L37

Commit:
Initial commit of mednist-ddpm

Commit:
mednist_ddpm commit. I have cleared the absolute paths to my personal directory.

Commit:
Added unit test to test the CXR sampling, which works now. Modified verify_bundle to pass the check for model.pt, since the requirement for two models (autoencoder and diffusion_model) makes sense for them to keep their specific names.
Modification of inference.json to add dummy attributes to pass the ConfigWorkflow check.
Modification of large_files.yml so that models are .pt and not .pth.

Commit:
Added unit test to test the CXR sampling, which works now. Modified verify_bundle to pass the check for model.pt, since the requirement for two models (autoencoder and diffusion_model) makes sense for them to keep their specific names.
Modification of inference.json to add dummy attributes to pass the ConfigWorkflow check.
Modification of large_files.yml so that models are .pt and not .pth.

Commit:
Removal of absolute path

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Run autofix

virginiafdez commented:
It still fails on the verify_bundle because I can only write one exception and there are two models that need to be downloaded that are not named model.pt.

Commit:
Modify script > inv function by operator.lt in train.yaml.

Add LICENSE file.

Commit:
Merge branch 'mednist_ddpm' of https://github.com/virginiafdez/model-zoo into mednist_ddpm

garciadias committed:
Merge remote-tracking branch 'origin/mednist_ddpm' into xray_classifier_port

Commit:
Change of autoencoder.pth to autoencoder.pt on scripts.
Modification of diffusion_model.pt name to model.pt to go through the verify_bundle Python function.

Commit:
Add authors to metadata.json.

Commit:
Remove mednist ddpm folder from the x-ray branch as it shouldn't be there.

Commit:
image feature missing from metadata.json. Added placeholder.

Commit:
Remove unecessary imports from test.

yiheng-wang-nv committed:
Merge branch 'dev' into xray_classifier_port

yiheng-wang-nv commented:
/build

KumoLiu commented:
```
[2025-01-21T02:40:11.254Z] 2025-01-21 02:40:11,230 - INFO - --- input summary of monai.bundle.scripts.verify_net_in_out ---
[2025-01-21T02:40:11.254Z] 2025-01-21 02:40:11,230 - INFO - > net_id: 'network_def'
[2025-01-21T02:40:11.254Z] 2025-01-21 02:40:11,230 - INFO - > meta_file: 'models/cxr_image_synthesis_latent_diffusion_model/configs/metadata.json'
[2025-01-21T02:40:11.254Z] 2025-01-21 02:40:11,230 - INFO - > config_file: 'models/cxr_image_synthesis_latent_diffusion_model/configs/inference.json'
[2025-01-21T02:40:11.254Z] 2025-01-21 02:40:11,230 - INFO - > bundle_root: 'models/cxr_image_synthesis_latent_diffusion_model'
[2025-01-21T02:40:11.254Z] 2025-01-21 02:40:11,230 - INFO - ---
[2025-01-21T02:40:11.254Z] 
[2025-01-21T02:40:11.254Z] 
[2025-01-21T02:40:11.510Z] Detected deprecated name 'optional_packages_version' in configuration file, replacing with 'required_packages_version'.
[2025-01-21T02:40:11.510Z] Traceback (most recent call last):
[2025-01-21T02:40:11.510Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/ci/verify_bundle.py", line 361, in <module>
[2025-01-21T02:40:11.510Z]     verify(bundle, models_path, mode)
[2025-01-21T02:40:11.510Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/ci/verify_bundle.py", line 342, in verify
[2025-01-21T02:40:11.510Z]     verify_data_shape(bundle_path, net_id, config_file)
[2025-01-21T02:40:11.510Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/ci/verify_bundle.py", line 196, in verify_data_shape
[2025-01-21T02:40:11.510Z]     verify_net_in_out(
[2025-01-21T02:40:11.510Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/model_zoo_venv/lib/python3.10/site-packages/monai/bundle/scripts.py", line 1230, in verify_net_in_out
[2025-01-21T02:40:11.510Z]     input_channels, input_spatial_shape, input_dtype, output_channels, output_dtype = _get_net_io_info(parser=parser)
[2025-01-21T02:40:11.510Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/model_zoo_venv/lib/python3.10/site-packages/monai/bundle/scripts.py", line 1132, in _get_net_io_info
[2025-01-21T02:40:11.510Z]     input_dtype = get_equivalent_dtype(parser.get(key), torch.Tensor)
[2025-01-21T02:40:11.510Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/model_zoo_venv/lib/python3.10/site-packages/monai/utils/type_conversion.py", line 85, in get_equivalent_dtype
[2025-01-21T02:40:11.510Z]     return dtype_numpy_to_torch(dtype)
[2025-01-21T02:40:11.510Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/model_zoo_venv/lib/python3.10/site-packages/monai/utils/type_conversion.py", line 64, in dtype_numpy_to_torch
[2025-01-21T02:40:11.510Z]     return torch.from_numpy(np.empty([], dtype=dtype)).dtype
[2025-01-21T02:40:11.510Z] TypeError: data type 'N/A' not understood
```
FYI @virginiafdez

ericspod commented on models/cxr_image_synthesis_latent_diffusion_model/configs/metadata.json:
```diff
@@ -0,0 +1,83 @@
+    "intended_use": "This is a research tool/prototype and not to be used clinically",
+    "network_data_format": {
+        "inputs": {
+            "image": {
+                "type": "N/A",
+                "format": "N/A",
+                "modality": "N/A",
+                "num_channels": 0,
+                "spatial_shape": [],
+                "dtype": "N/A",
+                "value_range": [],
+                "is_patch_data": false
+            },
```
Do we need this here? This should be specifying the input arguments for the diffusion model which is the "main" model for this bundle, so I don't think we need this argument at all.

virginiafdez replied:
I just removed it. The problem is that then, verify_bundle is complaining that this field "image" is required (along with the properties). 
Since it does not apply in this scenario because the model does not require images for inference I don't know what to put.

ericspod commented on models/cxr_image_synthesis_latent_diffusion_model/configs/inference.json:
```diff
@@ -0,0 +1,115 @@
+        "with_encoder_nonlocal_attn": false,
+        "with_decoder_nonlocal_attn": false
+    },
+    "network_def": "@autoencoder_def",
```
```suggestion
    "network_def": "@diffusion_def",
```
If the "main" network for this bundle is the diffusion model then this should be referred to here, though I don't know if it really matters for correct functioning of the bundle.

Commit:
Remove unecessary fields from metadata; change network def.

ericspod commented on models/cxr_image_synthesis_latent_diffusion_model/configs/metadata.json:
```diff
@@ -0,0 +1,73 @@
+{
+    "schema": "https://github.com/Project-MONAI/MONAI-extra-test-data/releases/download/0.8.1/meta_schema_20220324.json",
```
```suggestion
    "schema": "https://github.com/Project-MONAI/MONAI-extra-test-data/releases/download/0.8.1/meta_schema_20240725.json",
```
Let's try the current schema to see if this resolves the issues about the network arguments.

Commit:
Remove unecessary fields from metadata; change network def.

Commit:
Move contents of common.yaml to train and inference.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

KumoLiu commented:
/build

KumoLiu commented:
```
[2025-02-03T10:02:40.155Z] 2025-02-03 10:02:39,967 - INFO - Downloaded: models/cxr_image_synthesis_latent_diffusion_model/models/model.pt
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,917 - INFO - Verified 'model.pt', md5: c3fd4c8e38cd1d7250a8903cca935823.
[2025-02-03T10:02:42.046Z] directory is verified correctly.
[2025-02-03T10:02:42.046Z] version and changelog are verified correctly.
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,918 - INFO - --- input summary of monai.bundle.scripts.verify_metadata ---
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,918 - INFO - > meta_file: 'models/cxr_image_synthesis_latent_diffusion_model/configs/metadata.json'
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,918 - INFO - > filepath: 'models/cxr_image_synthesis_latent_diffusion_model/eval/schema.json'
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,918 - INFO - ---
[2025-02-03T10:02:42.046Z] 
[2025-02-03T10:02:42.046Z] 
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,919 - INFO - Expected md5 is None, skip md5 check for file models/cxr_image_synthesis_latent_diffusion_model/eval/schema.json.
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,919 - INFO - File exists: models/cxr_image_synthesis_latent_diffusion_model/eval/schema.json, skipped downloading.
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,942 - INFO - metadata is verified with no error.
[2025-02-03T10:02:42.046Z] metadata format is verified correctly.
[2025-02-03T10:02:42.046Z] properties are verified correctly.
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,942 - INFO - --- input summary of monai.bundle.scripts.verify_net_in_out ---
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,942 - INFO - > net_id: 'network_def'
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,942 - INFO - > meta_file: 'models/cxr_image_synthesis_latent_diffusion_model/configs/metadata.json'
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,943 - INFO - > config_file: 'models/cxr_image_synthesis_latent_diffusion_model/configs/inference.json'
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,943 - INFO - > bundle_root: 'models/cxr_image_synthesis_latent_diffusion_model'
[2025-02-03T10:02:42.046Z] 2025-02-03 10:02:41,943 - INFO - ---
[2025-02-03T10:02:42.046Z] 
[2025-02-03T10:02:42.046Z] 
[2025-02-03T10:02:42.046Z] Traceback (most recent call last):
[2025-02-03T10:02:42.046Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/ci/verify_bundle.py", line 361, in <module>
[2025-02-03T10:02:42.046Z]     verify(bundle, models_path, mode)
[2025-02-03T10:02:42.046Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/ci/verify_bundle.py", line 342, in verify
[2025-02-03T10:02:42.046Z]     verify_data_shape(bundle_path, net_id, config_file)
[2025-02-03T10:02:42.046Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/ci/verify_bundle.py", line 196, in verify_data_shape
[2025-02-03T10:02:42.046Z]     verify_net_in_out(
[2025-02-03T10:02:42.046Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/model_zoo_venv/lib/python3.10/site-packages/monai/bundle/scripts.py", line 1230, in verify_net_in_out
[2025-02-03T10:02:42.046Z]     input_channels, input_spatial_shape, input_dtype, output_channels, output_dtype = _get_net_io_info(parser=parser)
[2025-02-03T10:02:42.046Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/model_zoo_venv/lib/python3.10/site-packages/monai/bundle/scripts.py", line 1130, in _get_net_io_info
[2025-02-03T10:02:42.046Z]     input_spatial_shape = tuple(parser.get(key))
[2025-02-03T10:02:42.046Z] TypeError: 'NoneType' object is not iterable
```
cc @virginiafdez

ericspod commented on models/cxr_image_synthesis_latent_diffusion_model/large_files.yml:
```diff
@@ -0,0 +1,9 @@
+large_files:
+  - path: "models/autoencoder.pt"
+    url: "https://drive.google.com/uc?export=download&id=1paDN1m-Q_Oy8d_BanPkRTi3RlNB_Sv_h"
+    hash_val: "7f579cb789597db7bb5de1488f54bc6c"
+    hash_type: "md5"
+  - path: "models/model.pt"
+    url: "https://drive.google.com/uc?export=download&id=1CjcmiPu5_QWr-f7wDJsXrCCcVeczneGT"
```
```suggestion
    url: "https://drive.google.com/uc?id=1paDN1m-Q_Oy8d_BanPkRTi3RlNB_Sv_h"
    hash_val: "7f579cb789597db7bb5de1488f54bc6c"
    hash_type: "md5"
  - path: "models/model.pt"
    url: "https://drive.google.com/uc?id=1CjcmiPu5_QWr-f7wDJsXrCCcVeczneGT"
```

I think one issue is that the URLs are malformed. Gdown will pull down the html page saying something was wrong rather than error out. You can try the URLs on the command line with `gdown` then check their hashes with `md5sum`.

ericspod commented:
> ```
> ...
> [2025-02-03T10:02:42.046Z]     verify_net_in_out(
> [2025-02-03T10:02:42.046Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/model_zoo_venv/lib/python3.10/site-packages/monai/bundle/scripts.py", line 1230, in verify_net_in_out
> [2025-02-03T10:02:42.046Z]     input_channels, input_spatial_shape, input_dtype, output_channels, output_dtype = _get_net_io_info(parser=parser)
> [2025-02-03T10:02:42.046Z]   File "/home/jenkins/agent/workspace/MONAI-zoo-premerge/model_zoo_venv/lib/python3.10/site-packages/monai/bundle/scripts.py", line 1130, in _get_net_io_info
> [2025-02-03T10:02:42.046Z]     input_spatial_shape = tuple(parser.get(key))
> [2025-02-03T10:02:42.046Z] TypeError: 'NoneType' object is not iterable
> ```
> 
> cc @virginiafdez

The issue here is the verification of network shape assumes there's an "image" input for the network, this needs to be updated later [here](https://github.com/Project-MONAI/MONAI/blob/8dcb9dc0e594059d87ce9882c20fe5a59340a6b2/monai/bundle/scripts.py#L1121) but for now we should add the exception for this bundle.

Commit:
Add cxr bundle to exception list in bundle_custom_data.

Commit:
Run autofix.

Commit:
Update isort version

ericspod commented on requirements-dev.txt:
```diff
@@ -9,7 +9,7 @@ pep8-naming
 pycodestyle
 pyflakes
 black
```
```suggestion
black<25
```
Try this to fix the black version. It's complaining about needing to reformat code that isn't in this proposed bundle and we don't want to modify that right now.

Commit:
Update black version

KumoLiu commented:
/build
